### PR TITLE
[usage] Implement BillingService

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

Implements `UpdateInvoices` RPC on `BillingService`.

There are 2 implementations:
1. NoOp implementation for Self-hosted
2. Stripe implementation which re-uses the existing `stripe.Client`

With this PR, the flow changes to the following:
1. Usage Controller is triggered with a cron
2. Usage is retrieved and stored in the DB
3. Instead of then triggering the old `BillingController`, the `BillingService.UpdateInvoices(sessions)` is invoked over a grpc self-connection

Behaviour of the component remains unchanged. Tests are updated to move to the corresponding parts.

The benefit of using na RPC is that we get metrics for the RPCs but also we incrementally decouple the UsageController from the Billing one. In subsequent PRs, a similar RPC will be introduced for collecting Usage and these two RPCs will then be chained. We started with Billing as it's easier to do initially and reduces the size of subsequent PRs. If you want to see how this looks, [here's a PoC](https://github.com/gitpod-io/gitpod/pull/11631) which converts both.

## Related Issue(s)

Relates to https://github.com/gitpod-io/gitpod/issues/10785

## How to test

Unit tests
Preview

## Release Notes

```release-note
NONE
```

## Werft options:

- [x] /werft with-preview with-billing

### CODE REVIEW

**Comment by roboquat** [comment]
@easyCZ: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/usage/pkg/apiv1/billing.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		return nil, status.Errorf(codes.InvalidArgument, "failed to compute credit summary")
+	}
+
+	err = s.stripeClient.UpdateUsage(credits)
+	if err != nil {
+		log.Log.WithError(err).Errorf("Failed to update stripe invoices.")
+		return nil, status.Errorf(codes.Internal, "failed to update stripe invoices")
+	}
+
+	return &v1.UpdateInvoicesResponse{}, nil
```
Comment (andrew-farries) [inline components/usage/pkg/apiv1/billing.go:44]: Is there something useful we can return here? The number of invoices updated perhaps?

File: components/usage/pkg/apiv1/billing_test.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+func TestCreditSummaryForTeams(t *testing.T) {
+	teamID := uuid.New().String()
+	teamAttributionID := db.NewTeamAttributionID(teamID)
+
+	scenarios := []struct {
+		Name     string
+		Sessions []*v1.BilledSession
+		Expected map[string]int64
```
Comment (andrew-farries) [inline components/usage/pkg/apiv1/billing_test.go:22]: Should we add a test with two different team attribution ids?

File: components/usage/pkg/controller/reconciler.go [hunk 3]
```diff
@@ -148,31 +152,36 @@ func instancesToUsageRecords(instances []db.WorkspaceInstanceForUsage, pricer *W
...
 
-		creditsPerTeamID[id] += int64(instance.CreditsUsed)
-	}
-
-	for teamID, credits := range creditsPerTeamID {
-		creditsPerTeamID[teamID] = int64(math.Ceil(float64(credits)))
+		sessions = append(sessions, &v1.BilledSession{
+			AttributionId:  string(instance.AttributionID),
+			UserId:         instance.UserID.String(),
+			TeamId:         "",
```
[the remark this answers is not in the document]
Comment (easyCZ) [reply components/usage/pkg/controller/reconciler.go:168]: We don't yet have the TeamID from the database. I've made it empty here to make that explicit when someone is looking into why TeamId is empty.

File: components/usage/pkg/apiv1/billing.go (hunk 1 above)
Comment (easyCZ) [reply components/usage/pkg/apiv1/billing.go:44]: That response is open for extension. For now, I'd keep it empty until we know what we need to return, then we can add

File: components/usage/pkg/apiv1/billing_test.go (hunk 2 above)
Comment (easyCZ) [reply components/usage/pkg/apiv1/billing_test.go:22]: Makes sense. Originally this test is just moved from the old location but it makes sense to ensure more variance.

**Review by andrew-farries — APPROVED** [review]
/hold for one extra test

**Comment by easyCZ** [comment]
/hold
For me to validate one part of this change.

**Comment by andrew-farries** [comment]
FYI, this looks like it's failing to establish the gRPC connection to itself in staging:

```
{"config":{"controllerSchedule":"1h","creditsPerMinuteByWorkspaceClass":{"default":0.1666666667,"gitpodio-internal-xl":0.3333333333},"stripeCredentialsFile":"stripe-secret/apikeys","contentServiceAddress":"content-service:8080","server":{"services":{"grpc":{"address":":9001"}}}},"level":"info","message":"Starting usage component.","serviceContext":{"service":"usage","version":"commit-c4708b30483eae2bcd07144178252291ea97c326"},"severity":"INFO","time":"2022-07-27T09:32:20Z"}
{"level":"info","message":"Starting usage controller.","serviceContext":{"service":"usage","version":"commit-c4708b30483eae2bcd07144178252291ea97c326"},"severity":"INFO","time":"2022-07-27T09:32:20Z"}
{"component":"grpc","level":"warning","message":"2022/07/27 09:32:20 WARNING: [core] [Channel #2 SubChannel #3] grpc: addrConn.createTransport failed to connect to {","serviceContext":{"service":"usage","version":"commit-c4708b30483eae2bcd07144178252291ea97c326"},"severity":"WARNING","time":"2022-07-27T09:32:20Z"}
{"component":"grpc","level":"warning","message":"  \"Addr\": \":9001\",","serviceContext":{"service":"usage","version":"commit-c4708b30483eae2bcd07144178252291ea97c326"},"severity":"WARNING","time":"2022-07-27T09:32:20Z"}
{"component":"grpc","level":"warning","message":"  \"ServerName\": \"localhost:9001\",","serviceContext":{"service":"usage","version":"commit-c4708b30483eae2bcd07144178252291ea97c326"},"severity":"WARNING","time":"2022-07-27T09:32:20Z"}
{"component":"grpc","level":"warning","message":"  \"Attributes\": null,","serviceContext":{"service":"usage","version":"commit-c4708b30483eae2bcd07144178252291ea97c326"},"severity":"WARNING","time":"2022-07-27T09:32:20Z"}
{"component":"grpc","level":"warning","message":"  \"BalancerAttributes\": null,","serviceContext":{"service":"usage","version":"commit-c4708b30483eae2bcd07144178252291ea97c326"},"severity":"WARNING","time":"2022-07-27T09:32:20Z"}
{"component":"grpc","level":"warning","message":"  \"Type\": 0,","serviceContext":{"service":"usage","version":"commit-c4708b30483eae2bcd07144178252291ea97c326"},"severity":"WARNING","time":"2022-07-27T09:32:20Z"}
{"component":"grpc","level":"warning","message":"  \"Metadata\": null","serviceContext":{"service":"usage","version":"commit-c4708b30483eae2bcd07144178252291ea97c326"},"severity":"WARNING","time":"2022-07-27T09:32:20Z"}
{"component":"grpc","level":"warning","message":"}. Err: connection error: desc = \"transport: Error while dialing dial tcp :9001: connect: connection refused\"","serviceContext":{"service":"usage","version":"commit-c4708b30483eae2bcd07144178252291ea97c326"},"severity":"WARNING","time":"2022-07-27T09:32:20Z"}
```

**Comment by andrew-farries** [comment]
although, after restarting the pod it seems fine 🤷

**Comment by easyCZ** [comment]
Will take a closer look. It's possible it's a race condition between the server starting and the connection.

**Comment by andrew-farries** [comment]
I made https://github.com/gitpod-io/gitpod/issues/11743 to track the race at startup.

**Pull request merged**